src/Tags.elm

```diff
@@ -0,0 +1,217 @@
+        |> Rule.withModuleVisitor (Rule.withSimpleModuleDefinitionVisitor (always []))
+        |> Rule.withModuleContextUsingContextCreator
+            { fromModuleToProject = fromModuleToProject
+            , fromProjectToModule = fromProjectToModule
+            , foldProjectContexts = Set.union
+            }
+        |> Rule.withDataExtractor dataExtractor
+        |> Rule.fromProjectRuleSchema
+
+
+commonTags : Set Tag
+commonTags =
+    Set.fromList
+        [ "paradigm:functional"
+        , "technique:immutability"
+        , "uses:module"
+        ]
+
+
+fromProjectToModule : Rule.ContextCreator ProjectContext ModuleContext
+fromProjectToModule =
+    Rule.initContextCreator
+        (\lookupTable _ ->
+            { lookupTable = lookupTable
+            , tags = Set.empty
+            }
+        )
+        |> Rule.withModuleNameLookupTable
+
+
+fromModuleToProject : Rule.ContextCreator ModuleContext ProjectContext
+fromModuleToProject =
+    Rule.initContextCreator (\{ tags } -> tags)
+
+
+dataExtractor : ProjectContext -> Value
+dataExtractor =
+    Set.toList >> Encode.list Encode.string
+
+
+expressionTagsRule : Rule
+expressionTagsRule =
+    Rule.newProjectRuleSchema "Tags" Set.empty
+        |> Rule.withModuleVisitor (Rule.withExpressionEnterVisitor expressionVisitor)
+        |> Rule.withModuleContextUsingContextCreator
+            { fromModuleToProject = fromModuleToProject
+            , fromProjectToModule = fromProjectToModule
+            , foldProjectContexts = Set.union
+            }
+        |> Rule.withDataExtractor dataExtractor
+        |> Rule.fromProjectRuleSchema
+
+
+expressionVisitor : Node Expression -> ModuleContext -> ( List never, ModuleContext )
+expressionVisitor ((Node range expression) as node) ({ lookupTable, tags } as context) =
+    case expression of
+        FunctionOrValue _ name ->
+            case LookupTable.moduleNameFor lookupTable node of
+                Nothing ->
+                    ( [], { context | tags = Set.union tags (matchExpression node) } )
+
+                Just originalModuleName ->
+                    ( [], { context | tags = Set.union tags (matchExpression (Node range (FunctionOrValue originalModuleName name))) } )
+
+        _ ->
+            ( [], { context | tags = Set.union tags (matchExpression node) } )
+
+
+matchExpression : Node Expression -> Set Tag
+matchExpression (Node range expression) =
+    case expression of
+        FunctionOrValue [ "Set" ] _ ->
+            Set.fromList [ "construct:set", "technique:immutable-collection", "technique:sorted-collection" ]
+
+        FunctionOrValue [ "Dict" ] _ ->
+            Set.fromList [ "construct:dictionary", "technique:immutable-collection", "technique:sorted-collection" ]
+
+        FunctionOrValue [ "Random" ] _ ->
+            Set.singleton "technique:randomness"
+
+        FunctionOrValue [ "Regex" ] _ ->
+            Set.singleton "technique:regular-expression"
+
+        FunctionOrValue [ "Debug" ] _ ->
```

Comment:
I think we forbid this in the Analyser elsewhere, as part of the common rules? Might be worth having here just in case though.

Reply:
That rule is `Actionable`, so I think it's not blocking students from publishing a solution with a debug anyway. Also solutions published before the rule would have them.

BTW, that just made me think of adding a `Debug` concept, that might be worth doing.